sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala

```diff
@@ -293,12 +295,16 @@ object RewriteDistinctAggregates extends Rule[LogicalPlan] {
           val operators = expressions.map { e =>
             val af = e.aggregateFunction
             val naf = patchAggregateFunctionChildren(af) { x =>
-              val condition = if (e.filter.isDefined) {
-                e.filter.map(distinctAggFilterAttrLookup.get(_)).get
+              val condition = e.filter.map(distinctAggFilterAttrLookup.get(_)).flatten
+              if (distinctAggGroupLookup(e).contains(x)) {
```

Comment:
I'm wondering if we can simplify the logic a bit. The goal is to only do the replacement for the first child if all the children are foldable. How about
```
val af = e.aggregateFunction
if (af.children.forall(_.foldable)) {
  val firstChild = evalWithinGroup(id, af.children.head, condition)
  af.withNewChildren(firstChild +: af.children.drop(1)).asInstanceOf[AggregateFunction]
} else {
  // the original code
}
```

Then we don't need to change a lot of code to create `distinctAggGroupLookup`

Reply:
OK